### ISSUE TITLE
annotations turn grey or go missing when executing "validate" from context menu

### ISSUE THREAD

**Issue by asheiduk** [eclipse/xtext-eclipse#1463]
The error/warning annotations in an open editor go missing or are greyed when explizit validation is triggered.

Steps to reproduce:
 - Use the "Hello World" project and enable the check.
 - Start guest eclipse.
 - Type a file like this:
```
        Hello Xtext!
        Hello foo!
```
- Save
- Notice that there is one warning annotation both in the left vertical ruler and in the right overview ruler.
- Open the context menu of the **editor**, select "Validate" (the one with the hover "Trigger expensive validation", not the other one) 
- Notice that the vertical ruler now contains only a grey warning triangle and the overview ruler is empty. The "Problems" view still contains the warning.

Further observation: It seems only "dirty" editors show this bug:

- Close the editor window and reopen it. The annotations are displayed.
- Select "Validate" again from the editor's context menu
- Now both annotations are still displayed.
- Type something and save
- Select "Validation"
- Notice that the annotations are missing again.

We see this behaviour in Xtext 2.14 (which my project uses right now.) but als in 2.21.

While debugging we came to `org.eclipse.xtext.ui.editor.validation.MarkerIssueProcessor` which seems to remove the markers/annotations and a reference to #462 which worked in the same area.

**Comment by miklossy** [comment]
Thanks for your bug report @asheiduk ! I could also reproduce the problem with Xtext 2.22.0.v20200525-0122
![screencast](https://user-images.githubusercontent.com/20393472/82904688-19b39680-9f63-11ea-8e41-d22dbb224e8d.gif)

**Comment by szarnekow** [comment]
Please note that "expensive" validations and the manual validate action are not meant to be used along with the XtextBuilder in the same project. If you intend to do that, the marker management is up to the language implementor.

**Comment by asheiduk** [comment]
@szarnekow I must admitt that I don't understand what you are trying to tell me.

The hover says "expensive Validation" but it executes **all** checks and therefore
handles **all** errors.

Both the menu action and the build command are configured by default.
So I guess everyone assumes that this **should** work together out of the box. 

I have no idea what **project specific** code would be required to align
them both. I guess that code would be better handled by Xtext directly.

**Comment by cdietrich** [comment]
ping @szarnekow

**Comment by cdietrich** [comment]
we are affected  by a similar issue.
debugging
{{org.eclipse.ui.texteditor.ResourceMarkerAnnotationModel.batchedUpdate(IMarkerDelta[])}}
i can see adds+deletes instead of modify
maybe thats the problem.

**Comment by cdietrich** [comment]
nope:
org.eclipse.xtext.ui.editor.validation.AnnotationIssueProcessor.updateMarkerAnnotations(IProgressMonitor)
marks them as gray as it assumes they will be recreated by the reconciler

**Comment by cdietrich** [comment]
maybe we can do something like

```
@Override
	public Object execute(ExecutionEvent event) throws ExecutionException {
		XtextEditor xtextEditor = EditorUtils.getActiveXtextEditor(event);
		if (xtextEditor != null) {
			IXtextDocument xtextDocument = xtextEditor.getDocument();
			IResource resource = xtextEditor.getResource();
			if(resource != null) {
				IValidationIssueProcessor issueProcessor = new MarkerIssueProcessor(resource, xtextEditor.getInternalSourceViewer().getAnnotationModel(), markerCreator, markerTypeProvider);
				IValidationIssueProcessor issueProcessor2 = new AnnotationIssueProcessor(xtextDocument, xtextEditor.getInternalSourceViewer().getAnnotationModel(), issueResolutionProvider);
				ValidationJob validationJob = new ValidationJob(resourceValidator, xtextDocument, new IValidationIssueProcessor() {
					
					@Override
					public void processIssues(List<Issue> issues, IProgressMonitor monitor) {
						List<Issue> normalAndExpensive = new ArrayList<Issue>();
						List<Issue> fast = new ArrayList<Issue>();
						for (Issue i : issues) {
							if (CheckType.FAST == i.getType()) {
								fast.add(i);
							} else {
								normalAndExpensive.add(i);
							}
						}
						issueProcessor.processIssues(normalAndExpensive, monitor);
						issueProcessor2.processIssues(fast, monitor);
						
					}
				},
						CheckMode.ALL);
				validationJob.schedule();
			} else {
				IValidationIssueProcessor issueProcessor = new AnnotationIssueProcessor(xtextDocument, xtextEditor.getInternalSourceViewer().getAnnotationModel(), issueResolutionProvider);
				ValidationJob validationJob = new ValidationJob(resourceValidator, xtextDocument, issueProcessor,
						CheckMode.ALL);
				validationJob.schedule();
			}
		}
		return this;
	}
```
in the handler.
@asheiduk  could you give that a try?

**Comment by cdietrich** [comment]
ok would not create a fast resource master.
so maybe both

**Comment by cdietrich** [comment]
maybe something like

```
for (Issue i : issues) {
							if (CheckType.FAST == i.getType()) {
								fast.add(i);
								normalAndExpensive.add(i);
							} else {
								normalAndExpensive.add(i);
							}
						}
```

**Comment by cdietrich** [comment]
or even simpler

```@Override
					public void processIssues(List<Issue> issues, IProgressMonitor monitor) {
						List<Issue> fast = new ArrayList<Issue>();
						for (Issue i : issues) {
							if (CheckType.FAST == i.getType()) {
								fast.add(i);
							}
						}
						issueProcessor.processIssues(issues, monitor);
						issueProcessor2.processIssues(fast, monitor);
						
					}
```

**Comment by asheiduk** [comment]
This fixes the issue for us -- assuming the code goes to `Vermittlungsschwerpunkte` ;-) 

Just for reference: While retesting I discovered that this is all about `FAST` checks -- `NORMAL` and `EXPENSIVE` seemed to work already fine.

**Comment by cdietrich** [comment]
@szarnekow what do you think?

**Comment by szarnekow** [comment]
I'd be ok with changing the `ValidateActionHandler`

**Comment by cdietrich** [comment]
Fixed in 2.24 (due in december)
